agent/consul/leader.go

```diff
@@ -33,25 +33,41 @@ func (s *Server) monitorLeadership() {
 	// cleanup and to ensure we never run multiple leader loops.
 	raftNotifyCh := s.raftNotifyCh
 
+	// This wait group will ensure that only one leader loop will ever be
+	// running at any given time. We need to run that in a goroutine so we
+	// can look for other notifications (loss of leadership) which we pass
+	// along by closing the stopCh given to the current leader loop.
 	var wg sync.WaitGroup
 	var stopCh chan struct{}
 	for {
 		select {
 		case isLeader := <-raftNotifyCh:
 			if isLeader {
+				if stopCh != nil {
```

Comment:
This looks correct. I have three small nitpicks:

1. I'd call `defer wg.Done` out of habit in case the function becomes more complex
2. I'd name go func arg different (e.g. `ch`) to avoid accidental future confusion
3. I think writing this as a switch statement may help readability since it gets rid of some brackets

```go
			switch {
			case isLeader:
				if stopCh != nil {
					s.logger.Printf("[ERR] consul: attempted to start the leader loop while running")
					continue
				}

				stopCh = make(chan struct{})
				wg.Add(1)
				go func(ch chan struct{}) {
					defer wg.Done()
					s.leaderLoop(ch)
				}(stopCh)
				s.logger.Printf("[INFO] consul: cluster leadership acquired")

			default:
				if stopCh == nil {
					s.logger.Printf("[ERR] consul: attempted to stop the leader loop while not running")
					continue
				}

				s.logger.Printf("[DEBUG] consul: shutting down leader loop")
				close(stopCh)
				wg.Wait()
				stopCh = nil
				s.logger.Printf("[INFO] consul: cluster leadership lost")
			}
```

Reply:
Thanks @magiconair these are all good suggestions. PTAL at the last push.